### ISSUE TITLE
SlackBot.GetConversationByName (Too Many Requests)

### ISSUE THREAD

**Issue by Silvenga** [soxtoby/SlackNet#58]
When using SlackBot.GetConversationByName within a large org, Slack responds with Too Many Requests when the bot attempts to fetch all conversations.

This means, every time GetConversationByName is called, the same fetch loop occurs, and ultimate fails.

```
System.Net.Http.HttpRequestException: Response status code does not indicate success: 429 (Too Many Requests).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at SlackNet.Http.Execute[T](HttpRequestMessage requestMessage, Nullable`1 cancellationToken)
   at SlackNet.SlackApiClient.Get[T](String apiMethod, Dictionary`2 args, Nullable`1 cancellationToken)
   at SlackNet.Bot.SlackBot.FetchConversations()
   at SlackNet.Bot.SlackBot.GetConversations()
   at SlackNet.Bot.SlackBot.FindConversation(Func`2 predicate)
   at SlackNet.Bot.SlackBot.GetConversationByName(String conversationName)
```

I wonder if this request should be separate out into a couple of methods e.g.
- search public and private channels
- search im/mpim

**Comment by soxtoby** [comment]
I've added an automatic delay and retry on all requests when getting rate-limited - hopefully this solves your problem.

If it doesn't, then I may have to consider splitting conversations into channel/ims etc., but ideally I'd like to keep them all together, since that's kind of the point of the conversations API.

**Comment by Silvenga** [comment]
Thanks!

Yeah, looks like those changes will work. The simplicity of the conversation api definitely makes things easier.

**Comment by Silvenga** [comment]
@soxtoby sorry it so long, looks like this won't work, the retry needs to be higher then: https://github.com/soxtoby/SlackNet/blob/1cdaedf0fa74deb319131a3f1c8074b9811cfc27/SlackNet/Http.cs#L26-L31

Since the `requestMessage` can't be reused.

```
System.InvalidOperationException: The request message was already sent. Cannot send the same request message multiple times.
   at System.Net.Http.HttpClient.CheckRequestMessage(HttpRequestMessage request)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at SlackNet.Http.Execute[T](HttpRequestMessage requestMessage, Nullable`1 cancellationToken)
   at SlackNet.SlackApiClient.Get[T](String apiMethod, Dictionary`2 args, Nullable`1 cancellationToken)
   at SlackNet.Bot.SlackBot.FetchConversations()
   at SlackNet.Bot.SlackBot.GetConversations()
   at SlackNet.Bot.SlackBot.FindConversation(Func`2 predicate)
   at SlackNet.Bot.SlackBot.GetConversationByName(String conversationName)
```

**Comment by soxtoby** [comment]
Hmm I guess I'll need to pull it up to SlackApiClient. Not _too_ bad I suppose.

**Comment by soxtoby** [comment]
Should be fixed in v0.7.7 🤞

**Comment by Silvenga** [comment]
Again, I'm super slow 😿. The thread is completely blocked. It looks like what's happening in https://github.com/soxtoby/SlackNet/issues/60 is valid.

**Comment by soxtoby** [comment]
The paging issue should be fixed in v0.7.8. Third time's a charm, hopefully 😅.

**Comment by Silvenga** [comment]
😹 thanks for your effort.
